VACMS-3197 aliases part b

ethanteague opened the pull request:
## Description
See #3197 
This pr:
1. Forces LC content types to use the patterns created in pr 3204.
2. Unchecks the auto path alias toggle switch after lc node is published.

## Testing done
Behat / visual

## QA steps
1. As an `administrator`,  visit `/node/8296/edit` and visually verify that Url alias widget field reads `/resources-and-support-for-your-va-benefits-and-services`
2. Publish node, then visit `/node/8296/edit` as a `content_editor` and visually verify that Url alias widget not checked.



- [x] As an content_editor change title prior to publish.  Validate that the url changes.
Publish the node
- [x] As an content_editor change title after publish.  Validate that the url does not change.
- [x] As an admin change title and validate that the url does not change.
- [x] As an admin change the alias and validate that the url changes.
- [x]  As and editor edit the title and save the change with a status of draft or ready for review (can not publish).
- [x] as admin or content admin, publish the edit from above.

## Post deployment steps
 - [ ] CD to docroot.
 - [ ] Execute `drush scr ../scripts/content/VACMS-3197-bulk_update_lc_path_aliases-2020-10.php`
**Expected**:
  -  Should take less than a minute to complete the run.
  -  There should be ~95 nodes updated.   
  -  There will be at least 5 failures.
there are at least 5 nodes that will not be updated because they already have paths set.
![image](https://user-images.githubusercontent.com/5752113/97768043-ad7f2500-1af6-11eb-9935-9c6844f98fe8.png)
- [ ] Paste terminal output into comment in this PR.

swirtSJW commented on scripts/content/VACMS-3197-bulk_update_lc_path_aliases-2020-10.php:
```diff
@@ -0,0 +1,59 @@
+    Drupal::logger('va_gov_db')->log(LogLevel::INFO, 'Updated alias for nid %nid completed by va_gov_db_update_8020.', [
+      '%nid' => $entity->id(),
+    ]);
+    $updated = $updated + 1;
```
Minor:
```suggestion
    $updated++;
```

swirtSJW commented on scripts/content/VACMS-3197-bulk_update_lc_path_aliases-2020-10.php:
```diff
@@ -0,0 +1,59 @@
+    $updated = $updated + 1;
+  }
+  else {
+    $failed = $failed + 1;
```
Minor:
```suggestion
    $failed++;
```

ethanteague replied:
Yeah, agreed - was copying stuff and not thinking much. Will update, @swirtSJW

swirtSJW reviewed (changes requested):
A couple changes please.

ethanteague commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -183,6 +184,28 @@ function _va_gov_backend_add_vamc_regions_select(array &$form) {
  * Implements hook_form_alter().
  */
 function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+
+  // Disable path alias modification after lc node is published.
+  if ($form_state->getFormObject() instanceof EntityFormInterface) {
+    $lc_bundles = [
+      'basic_landing_page',
+      'checklist',
+      'faq_multiple_q_a',
+      'media_list_images',
+      'media_list_videos',
+      'q_a',
+      'step_by_step',
+    ];
+    $bundle = $form_state->getformObject()->getEntity()->bundle();
+    if (in_array($bundle, $lc_bundles)) {
+      $status = $form_state->getformObject()->getEntity()->get('moderation_state')->getString();
+      // If it's published and the user isn't an admin,
+      // the alias widget is not available.
+      if ($status === 'published' && !in_array('administrator', \Drupal::currentUser()->getRoles())) {
+        $form['path']['#access'] = FALSE;
+      }
+    }
+  }
```
Yep - I got lazy. Will update, @swirtSJW :+1:

swirtSJW reviewed (commented):
I understand this takes care of hiding the field if it is published.  Is there are part C that contains:

 1. the entity_update hook that  unticks the checkbox when it is changed from draft to published.  
2. script to change tick the box on all currently published nodes.

ethanteague commented:
> I understand this takes care of hiding the field if it is published. Is there are part C that contains:
> 
> 1. the entity_update hook that  unticks the checkbox when it is changed from draft to published.
> 2. script to change tick the box on all currently published nodes.

@swirtSJW @oksana-c I had a tough time digging into the widget to alter that checkbox. I was able to pretty easily just remove the whole widget after publish, so that's what I'm doing in this pr. So, the tick box logic isn't needed, because the box is unavailable after publish to non-admins (it returns if the node is set back to draft / review). Hoping you all are okay with this approach - I couldn't seem to dig into that checkbox :cry:

swirtSJW commented:
> So, the tick box logic isn't needed, because the box is unavailable after publish to non-admins (it returns if the node is set back to draft / review). Hoping you all are okay with this approach - I couldn't seem to dig into that checkbox

I think it is still needed.  I don't think hiding the tick box  preventz the node from still following the url pattern as a title changes.   We need it to actually remain with the current url indefinitely.

ethanteague commented:
> > So, the tick box logic isn't needed, because the box is unavailable after publish to non-admins (it returns if the node is set back to draft / review). Hoping you all are okay with this approach - I couldn't seem to dig into that checkbox
> 
> I think it is still needed. I don't think hiding the tick box prevents the node from still following the url pattern as a title changes. We need it to actually remain with the current url indefinitely.

Yeah, you are right. Updated, @swirtSJW :+1:

swirtSJW commented:
On testing this out @ethanteague  I see an issue.

If as an admin I publish a checklist, then edit the checklists title and save the node, I get this error

![image](https://user-images.githubusercontent.com/5752113/97344771-9bd02000-185f-11eb-8d41-f6bce78e3601.png)

swirtSJW reviewed (changes requested):
Solve issue reported in comment related to node save error.

ethanteague commented:
> On testing this out @ethanteague I see an issue.
> 
> If as an admin I publish a checklist, then edit the checklists title and save the node, I get this error
> 
> ![image](https://user-images.githubusercontent.com/5752113/97344771-9bd02000-185f-11eb-8d41-f6bce78e3601.png)
> 
> The save never happens.

> On testing this out @ethanteague I see an issue.
> 
> If as an admin I publish a checklist, then edit the checklists title and save the node, I get this error
> 
> ![image](https://user-images.githubusercontent.com/5752113/97344771-9bd02000-185f-11eb-8d41-f6bce78e3601.png)
> 
> The save never happens.

Ugh....my bad! I just pushed a fix, @swirtSJW

swirtSJW commented:
@ethanteague  This is mainly working correctly, but has an issue.
If a node gets published the checkmark goes away as it should.   But then if an editor who can only edit and submit drafts for review, makes a title change and then submits for review, and then I as an admin looks at the draft to change it to published, the checkmark is again set.
![image](https://user-images.githubusercontent.com/5752113/97466159-2cb70200-1919-11eb-8c01-f59c3ebebed3.png)

When I publish the revision, the node gets and updated alias that should not have happened.
![image](https://user-images.githubusercontent.com/5752113/97466409-73a4f780-1919-11eb-81d2-22284588b4ab.png)

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,36 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+      $status = $form_state->getformObject()->getEntity()->get('moderation_state')->getString();
+      // If it's published the alias widget is unchecked
+      // to disallow auto path updating.
+      if ($status === 'published') {
```
I think only having this look at published means the it does not untick the box when you are looking at  node that has been published but the current revision that you are looking at is in review or draft.   It has to look at node itself, not the revision.

ethanteague replied:
Yep, that's the ticket, @swirtSJW - updated :+1:

swirtSJW reviewed (changes requested):
This is largely working, but has one flaw.   See error described in previous comment.

ethanteague commented:
@swirtSJW - reworked this a bit. I've gotten rid of the js checkbox toggle approach, and replaced with an after_build function. I've also added a check in entity presave hook to turn off pathauto if the default revision is published. Hopefully this will cover us! Appreciate your eyes on this.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,62 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+    'media_list_images',
+    'media_list_videos',
+    'q_a',
+    'step_by_step',
```
```suggestion
    'step_by_step',
    'va_form',
```

This feature was originally requested for VA Forms nodes.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,62 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
- * Implements hook_field_group_build_pre_render_alter().
+ * Returns the R&S content types.
+ *
+ * Return an array of R&S node types.
```
```suggestion
 * @return array 
 *   Node types that should have their alias locked upon publishing.
```

@return?     I think your codesniffer be stuffed up.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,62 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+ *
+ * Return an array of R&S node types.
+ */
+function _va_gov_backend_r_and_s_types() {
```
```suggestion
function _va_gov_backend_get_locked_alias_bundles() {
```

Functions should be verbs, and variables should be nouns. Also I think the name of this should be less product specific and more about what it represents. Something like "_va_gov_backend_get_locked_alias_bundles"

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,62 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+ */
+function _va_gov_disable_alias_widget(array &$form, FormStateInterface $form_state) {
+  if ($form_state->getFormObject() instanceof EntityFormInterface) {
+    $bundles_with_hidden_alias_widget = _va_gov_backend_r_and_s_types();
```
```suggestion

```

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,62 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+  if ($form_state->getFormObject() instanceof EntityFormInterface) {
+    $bundles_with_hidden_alias_widget = _va_gov_backend_r_and_s_types();
+    $bundle = $form_state->getformObject()->getEntity()->bundle();
+    if (in_array($bundle, $bundles_with_hidden_alias_widget)) {
```
```suggestion
    if (in_array($bundle, _va_gov_backend_get_locked_alias_bundles())) {
```

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -1111,13 +1167,34 @@ function va_gov_backend_pathauto_alias_alter(&$alias, array &$context) {
     _va_gov_backend_sync_wysiwyg_fields($entity);
   }
+
+  if ($entity->getEntityTypeId() === 'node' && in_array($entity->bundle(), $r_and_s_types)) {
```
```suggestion
  if ($entity->getEntityTypeId() === 'node' && in_array($entity->bundle(), _va_gov_backend_get_locked_alias_bundles())) {
```

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -1111,13 +1167,34 @@ function va_gov_backend_pathauto_alias_alter(&$alias, array &$context) {
 function va_gov_backend_entity_presave(EntityInterface $entity) {
   $bundle_types = ['event', 'news_story', 'press_release', 'outreach_asset'];
 
+  $r_and_s_types = _va_gov_backend_r_and_s_types();
+
```
```suggestion
```

swirtSJW reviewed (changes requested):
I added some suggestion.

ethanteague commented:
> I added some suggestion.

Thanks @swirtSJW - added them.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,63 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+ * @param array $form
+ *   The node form array.
+ *
+ *   return array $form.
```
```suggestion
 *   @return array 
  *   The $form.
```

code sniffer

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -303,11 +308,64 @@ function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form
+ * @param Drupal\Core\Form\FormStateInterface $form_state
+ *   Instance of FormStateInterface.
+ */
+function _va_gov_disable_alias_widget(array &$form, FormStateInterface $form_state) {
+  if ($form_state->getFormObject() instanceof EntityFormInterface) {
+    $bundle = $form_state->getformObject()->getEntity()->bundle();
+    if (in_array($bundle, _va_gov_backend_get_locked_alias_bundles())) {
+      $entity_id = $form_state->getformObject()->getEntity()->id();
+      if (!empty($entity_id)) {
+        $entity = \Drupal::entityTypeManager()->getStorage('node')->load($entity_id);
+        // If it's published the alias widget is unchecked
+        // to disallow auto path updating.
+        if ($entity->isPublished() === TRUE) {
+          $form['#after_build'][] = '_va_gov_backend_node_after_build_uncheck_auto_alias';
+        }
+      }
+    }
+  }
+}
```
I don't think this is actually doing anything.   When I am content-editor I can never see the widget.   And when I am content-admin I can always see it, even though we are not hiding it based on role at all.

I am digging into this.

swirtSJW commented:
Testing conversation here just for history.
https://dsva.slack.com/archives/CT4GZBM8F/p1604065580014100

olivereri commented:
```
bash-4.2$ source /etc/sysconfig/httpd; /usr/local/bin/drush scr ../scripts/content/VACMS-3197-bulk_update_lc_path_aliases-2020-10.php 
Skipped or failed updating alias for nid 9275 VACMS-3197-bulk_update_lc_path_aliases-2020-10.                                                                                                       [warning]
Skipped or failed updating alias for nid 9480 VACMS-3197-bulk_update_lc_path_aliases-2020-10.                                                                                                       [warning]
Skipped or failed updating alias for nid 9276 VACMS-3197-bulk_update_lc_path_aliases-2020-10.                                                                                                       [warning]
Skipped or failed updating alias for nid 8684 VACMS-3197-bulk_update_lc_path_aliases-2020-10.                                                                                                       [warning]
Path aliases updated 84 out of 88 entities.  Skipped or failed to update: 4.
bash-4.2$
```
